src/field_theory/chevalley_warning.lean

```diff
@@ -0,0 +1,195 @@
+
+This file contains a proof of the Chevalley–Warning theorem.
+Throughout most of this file, `K` denotes a finite field
+and `q` is notation for the cardinality of `K`.
+
+## Main results
+
+1. Let `f` be a multivariate polynomial in finitely many variables (`X s`, `s : σ`)
+   such that the total degree of `f` is less than `(q-1)` times the cardinality of `σ`.
+   Then the evaluation of `f` on all points of `σ → K` (aka `K^σ`) sums to `0`.
+   (`sum_mv_polynomial_eq_zero`)
+2. The Chevalley–Warning theorem (`char_dvd_card_solutions`).
+   Let `f i` be a finite family of multivariate polynomials
+   in finitely many variables (`X s`, `s : σ`) such that
+   the sum of the total degrees of the `f i` is less than the cardinality of `σ`.
+   Then the number of common solutions of the `f i`
+   is divisible by the characteristic of `K`.
+
+## Notation
+
+- `K` is a finite field
+- `q` is notation for the cardinality of `K`
+- `σ` is the indexing type for the variables of a multivariate polynomial ring over `K`
+
+-/
+
+universes u v
+
+open_locale big_operators
+
+section finite_field
+open mv_polynomial function finset finite_field
+
+variables {K : Type*} {σ : Type*} [fintype K] [field K] [fintype σ]
+local notation `q` := fintype.card K
+
+lemma mv_polynomial.exists_degree_lt_card_sub_one (f : mv_polynomial σ K)
+  (h : f.total_degree < (q - 1) * fintype.card σ) {d : σ →₀ ℕ} (hd : d ∈ f.support) :
+  ∃ i, d i < q - 1 :=
+begin
+  have hq : 0 < q - 1,
+  { rw [← card_units, fintype.card_pos_iff],
+    exact ⟨1⟩ },
+  contrapose! h,
+  have : d.support = univ,
+  { rw eq_univ_iff_forall,
+    intro i, rw [finsupp.mem_support_iff, ← nat.pos_iff_ne_zero],
+    exact lt_of_lt_of_le hq (h _) },
+  calc (q - 1) * fintype.card σ = ∑ s:σ, (q - 1) :
+                                  by rw [sum_const, nat.smul_eq_mul, mul_comm, card_univ]
+                            ... ≤ ∑ s, d s : sum_le_sum (λ s _, h s)
+                            ... ≤ d.sum (λ (n : σ) (e : ℕ), e) : by rwa [finsupp.sum, this]
+                            ... ≤ f.total_degree : finset.le_sup hd,
+end
+
+lemma mv_polynomial.sum_mv_polynomial_eq_zero [decidable_eq σ] (f : mv_polynomial σ K)
+  (h : f.total_degree < (q - 1) * fintype.card σ) :
+  (∑ x, f.eval x) = 0 :=
+begin
+  haveI : decidable_eq K := classical.dec_eq K,
+  calc (∑ x, f.eval x) = ∑ x : σ → K, ∑ d in f.support, f.coeff d * ∏ i, x i ^ d i : _
+                   ... = ∑ d in f.support, ∑ x : σ → K, f.coeff d * ∏ i, x i ^ d i : sum_comm
+                   ... = 0 : sum_eq_zero _,
+  { simp only [eval, eval₂, finsupp.sum, id.def, finsupp.prod_pow], refl, },
+  intros d hd,
+  obtain ⟨i, hi⟩ : ∃ i, d i < q - 1, from f.exists_degree_lt_card_sub_one h hd,
+  let cd : K := f.coeff d,
+  calc (∑ x : σ → K, cd * ∏ i, x i ^ d i) = cd * (∑ x : σ → K, ∏ i, x i ^ d i) : mul_sum.symm
+                                      ... = 0 : (mul_eq_zero.mpr ∘ or.inr) _,
+  calc (∑ x : σ → K, ∏ i, x i ^ d i) =
```

Comment:
I think it's easier to read when `=` starts the next line. Then the `=` on line 84 can match its alignment. Same on line 87 and line 94.

Reply:
Done